contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,118 @@
+
+In Q2, we set a goal of "rock solid-batch processing" and [chose Temporal](https://github.com/PostHog/meta/pull/99) to build it because it helped solve issues like:
+
+- Retry, resume, cancel, and timeout logic per activity and workflow
+- Provides assurances about running exactly-once
+- Visibility into the status of workflows, including errors and logging.
+
+Instead of building all this functionality ourselves, Temporal takes care of it for us. After some debate, we felt building a workflow engine would not be a competitive advantage for us. Nothing about what we build on Temporal is novel enough to justify an investment into a competing runtime (as satisfying as it would be to write).
+
+## How we built our new export system with Temporal
+
+We didn’t completely rebuild our export system; many of the pieces still worked, we just needed to modify them to work with Temporal’s scheduling and execution capabilities. This meant building a batch export abstraction layer on top of Temporal. Specifically, the abstraction layer is four models:
+
+- `BatchExportDestination`: How users configure where to export data their data.
+- `BatchExportSchedule`: Maps to a [Temporal Schedule](https://keithtenzer.com/temporal/temporal_schedules/). It stores Schedule configurations to pass to Temporal.
+- `BatchExport`: The union of a `BatchExportSchedule` with a `BatchExportDestination`. Defines a destination for exported data, and a schedule to run said exports.
+- `BatchExportRun`: A `BatchExport` for each triggered workflow. Users aren’t meant to create these, only Temporal. They mostly serve to report on individual workflow executions.
+
+Users create a `BatchExport` that includes a schedule (which validates against the `BatchExportSchedule`) and a destination (stored as a `BatchExportDestination`). Data then gets read from PostHog and processed to write to the destination.
+
+With this structure in place, we began working on improving reliability, transparency, and performance.
+
+### Improving reliability and adding retries
+
+We prioritized improving reliability. As mentioned, Temporal has support for retry and timeout logic per activity and workflow which made a big difference. What we needed to do was tune this to our system and destinations. We needed to tune four main pieces:
+
+1. **ClickHouse** - [spaced out retries](https://github.com/PostHog/posthog/pull/15558) when reconnecting to prevent failures if in use by another query, retry [connection errors](https://github.com/PostHog/posthog/pull/16574) which are likely network failures rather than outages, use [offline ClickHouse cluster](https://github.com/PostHog/posthog/pull/16470) to improve performance and allowable query length, and sacrificed some performance to [de-duplicate batches](https://github.com/PostHog/posthog/pull/16354) by adding `DISTINCT ON` clause (because it handles queries asynchronously so de-duplication isn’t guaranteed).
```

Comment:
> because it handles queries asynchronously so de-duplication isn’t guaranteed

This reads a bit imprecise:

It's ClickHouse that de-duplicates asynchronously, so when querying (sync or async) de-duplicated is not guaranteed to have happened already.

```suggestion
1. **ClickHouse** - [spaced out retries](https://github.com/PostHog/posthog/pull/15558) when reconnecting to prevent failures if in use by another query, retry [connection errors](https://github.com/PostHog/posthog/pull/16574) which are likely network failures rather than outages, use [offline ClickHouse cluster](https://github.com/PostHog/posthog/pull/16470) to improve performance and allowable query length, and sacrificed some performance to [de-duplicate batches](https://github.com/PostHog/posthog/pull/16354) by adding `DISTINCT ON` clause (because ClickHouse handles de-duplication asynchronously so de-duplication isn’t guaranteed at query time).
```